pkg/ring/kv/dynamodb/client.go

```diff
@@ -0,0 +1,313 @@
+
+		if err != nil {
+			level.Error(c.logger).Log("msg", "error getting key", "key", key, "err", err)
+			continue
+		}
+
+		buf, err := c.codec.EncodeMultiKey(toUpdate)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "error serialising value", "key", key, "err", err)
+			continue
+		}
+
+		for childKey, bytes := range buf {
+			err := c.kv.Put(ctx, dynamodbKey{primaryKey: key, sortKey: childKey}, bytes)
+			if err != nil {
+				level.Error(c.logger).Log("msg", "error CASing", "key", key, "err", err)
+				continue
+			}
+		}
+
+		for _, childKey := range toDelete {
+			err := c.kv.Delete(ctx, dynamodbKey{primaryKey: key, sortKey: childKey})
+			if err != nil {
+				level.Error(c.logger).Log("msg", "error CASing", "key", key, "err", err)
+				continue
+			}
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("failed to CAS %s", key)
+}
+
+func (c *Client) WatchKey(ctx context.Context, key string, f func(interface{}) bool) {
+	bo := backoff.New(ctx, backoffConfig)
+
+	for bo.Ongoing() {
+		out, _, err := c.kv.Query(ctx, dynamodbKey{
+			primaryKey: key,
+		}, false)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "WatchKey", "key", key, "err", err)
+
+			if bo.NumRetries() > 10 {
+				if staleData := c.getStaleData(key); staleData != nil {
+					if !f(staleData) {
+						return
+					}
+					bo.Reset()
+				}
+			}
+			bo.Wait()
+			continue
+		}
+
+		decoded, err := c.decodeMultikey(out)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "error decoding key", "key", key, "err", err)
+			continue
+		}
+		c.updateStaleData(key, decoded)
+
+		if !f(decoded) {
+			return
+		}
+
+		bo.Reset()
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultLoopDelay):
+		}
+	}
+}
+
+func (c *Client) WatchPrefix(ctx context.Context, prefix string, f func(string, interface{}) bool) {
+	bo := backoff.New(ctx, backoffConfig)
+
+	for bo.Ongoing() {
+		out, _, err := c.kv.Query(ctx, dynamodbKey{
+			primaryKey: prefix,
+		}, true)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "WatchPrefix", "prefix", prefix, "err", err)
+			bo.Wait()
+			continue
+		}
+
+		for key, bytes := range out {
+			decoded, err := c.codec.Decode(bytes)
+			if err != nil {
+				level.Error(c.logger).Log("msg", "error decoding key", "key", key, "err", err)
+				continue
+			}
+			if !f(key, decoded) {
+				return
+			}
+		}
+
+		bo.Reset()
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultLoopDelay):
+		}
+	}
+}
+
+func (c *Client) decodeMultikey(data map[string][]byte) (codec.MultiKey, error) {
+	res, err := c.codec.DecodeMultiKey(data)
+	if err != nil {
+		return nil, err
+	}
+	out, ok := res.(codec.MultiKey)
+	if !ok || out == nil {
+		return nil, fmt.Errorf("invalid type: %T, expected MultikKey", out)
+	}
+
+	return out, nil
+}
+
+func (c *Client) LastUpdateTime(key string) time.Time {
+	c.staleDataLock.Lock()
+	defer c.staleDataLock.Unlock()
+
+	data, ok := c.staleData[key]
+	if !ok {
+		return time.Time{}
```

Comment:
Why return time `zero` here?
In the UI, I got this `Storage updated: 0001-01-01 00:00:00 +0000 UTC`

Reply:
My thought was return an empty date as we don't have any information stored yet. 
But this should not happen in the UI.
I had a look at it and I missed to update the prefix proxy that the ring have it. The key used by the function was different from the ring. I am changing the code to fix it. Thanks